text/0000-procedural-vtables.md

```diff
@@ -0,0 +1,338 @@
+By default, if no vtable generator function is specified for a specific trait,
+the unspecified scheme used today keeps getting used.
+
+# Motivation
+[motivation]: #motivation
+
+The only way we're going to satisfy all users' use cases is by allowing users
+complete freedom in how their wide pointers' metadata is built.
+Instead of hardcoding certain vtable layouts in the language
+(https://github.com/rust-lang/rfcs/pull/2955) we can give users the capability
+to invent their own wide pointers (and thus custom dynamically sized types).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In order to mark a trait (`MyTrait`) as using a custom vtable layout, you implement the `CustomUnsized` trait for `dyn MyTrait`.
+
+```rust
+trait MyTrait: SomeSuperTrait {
+    fn some_fn(&mut self, x: u32) -> i32;
+}
+
+impl CustomUnsized for dyn MyTrait {
+    type WidePointer = MyWidePointer;
+    // details later in this RFC
+}
+
+```
+
+`MyWidePointer` is the backing type that is going to be used for wide pointers to the given trait, so
+
+* `&dyn MyTrait`
+* `Box<dyn MyTrait>`
+* `Arc<dyn MyTrait>`
+
+and any other container types that can use wide pointers. Normally when unsizing from a concrete
+pointer like `&MyStruct` to `&dyn MyTrait` a wide pointer (that essentially is
+`(&MyStruct, &'static Vtable)`) is produced. This is currently done via compiler magic, but
+further down in this section you can see how it theoretically could be done in user code.
+Likely it will stay compiler magic out of compile-time performance reasons.
+
+All `impl`s of `MyTrait` will now use `MyWidePointer` for generating the wide pointer.
+The `TraitDescription` struct describes the metadata like the list of methods and a tree
+of super traits.
+
+You are actually generating the wide pointer, not a description of it.
+Since your `impl`'s `from` function is being interpreted in the target's environment, all target specific information will match up.
+Now, you need some information about the `impl` in order to generate your metadata (and your vtable).
+You get this information directly from the type (the `T` parameter) by adding trait bounds on it.
+
+As an example, consider the function which is what normally generates your metadata.
+
+```rust
+#[repr(C)]
+struct Pointer<T> {
+    ptr: *mut T,
+    vtable: &'static VTable<T>,
+}
+
+/// If the `owned` flag is `true`, this is an owned conversion like
+/// in `Box<T> as Box<dyn Trait>`. This distinction is important, as
+/// unsizing that creates a vtable in the same allocation as the object
+/// (like C++ does), cannot work on non-owned conversions. You can't just
+/// move away the owned object. The flag allows you to forbid such
+/// unsizings by triggering a compile-time `panic` with an explanation
+/// for the user.
+unsafe impl<T: MyTrait> const CustomUnsize<Pointer> for T {
+    fn unsize<const owned: bool>(ptr: *mut T) -> Pointer<T> {
```

Comment:
Idk, I haven't been able to come up with anything better so far

Reply:
I'm worried this doesn't actually work. `ptr` may be owned but not point to the start of the allocation. For example:
```rust
use core::ops::CoerceUnsized;

struct MyWrapper<T: ?Sized> {
    some_data: u32,
    object: T,
}

impl<T: CoerceUnsized<U>, U> CoerceUnsized<MyWrapper<U>> for MyWrapper<T> {}

fn bad() -> Box<MyWrapper<dyn MyTrait>> {
    Box::new(MyWrapper {
        some_data: 42,
        object: MyTraitImplementor,
    })
    // Undefined behavior: dealloc is called with a pointer 4 bytes offset from the allocation
} 
```

I think this might be solved this by making the `CustomUnsize` impl optional. I originally thought it had to be required but I suppose it should be possible for a user to customize the `dyn Trait` type via `CustomUnsized` *without* allowing safe `Box::new(TraitImplementor) as Box<dyn Trait>` conversion via `CustomUnsize`.  So to do the C++ shenanigans you would instead specialize the unsizing to only sound cases like so:
```rust
pub fn boxed_class<T: MyTrait>(data: T) -> Box<dyn MyTrait> {
    let ptr = Box::into_raw(Box::new((mytrait_vtable::<T>(), data)));
    unsafe { Box::from_raw(std::mem::transmute(ptr)) }
}
```

Reply:
For C++, I think you would need `Pin<Box<dyn MyTrait>>` since C++ assumes objects don't move.

Reply:
>  `ptr` may be owned but not point to the start of the allocation

The `CustomUnsize` impl gets a pointer to the thing being unsized, which in this case is `MyWrapper<i32>`, so this would all work out. The only time you don't have a pointer to the base is when you manually play with allocations, which may indeed be a problem (so if you have a custom `Box` that prepends with some padding).

Reply:
> The CustomUnsize impl gets a pointer to the thing being unsized

Ah, that explains my confusion around the slice example. I'll have to think on the implications of it a bit.